Package binary as a Python wheel

ejohb opened the pull request:
See #226

ejohb committed:
Makefile and setup.py updates for binary wheel

Modified the Makefile to add `setuptools` and `injectlib` targets, to add shared library. Plus add binary/wheel packaging.

ejohb committed:
Refactor library path resolution.

Updated the initialization logic to prioritize locating the shared library in the bindings directory, for binary wheel distributions.

ejohb committed:
Refactor setup.py and library handling.

Reorganized setup.py for better modularity, replacing hardcoded variables with dynamically retrieved values. Introduced `get_libyamlscript_ext` for determining library extensions

ejohb committed:
Further setup.py/__init__.py refactoring

ingydotnet commented:
The tricky part here is where I build and publish the wheels from.

I need to explain the release process. It is partially automated and a bit involved.

YAMLScript's `ys` and `libyamlscript.{so,dylib}` native binaries ship to 4 platforms: the matrix of linux/macos and intel/arm as seen here: https://github.com/yaml/yamlscript/releases

I do all the releases by hand currently. I have a linux/intel laptop and I have 3 other machines for the other 3 platforms that I can ssh into.

I run a command that looks like:
```
YS_GH_USER=ingydotnet YS_GH_TOKEN=<github-api-token> make release-yamlscript  o=0.1.96 n=0.1.97
```

This runs `util/release-yamlscript` a script written in yamlscript. It has 11 steps, and is pretty well commented. You should give it a read.

The first 9 steps build and release the native artifacts to a new https://github.com/yaml/yamlscript/releases version section (for linux/intel).
The script then pauses after printing the new github release id number.

At this point I ssh into the other 3 machines and run:
```
YS_GH_USER=ingydotnet YS_GH_TOKEN=<github-api-token> make release-yamlscript  RELEASE_ID=213416675
```

When the `RELEASE_ID` var is set, the script only runs step 9: `make release-assets`

Back on my linux/intel machine I press enter and the script continues with:

Step 10 does the release of each language binding (including to pypi).

Step 11 publishes the website.

Then it is complete.

----

For this we need to publish wheels somewhere in here. I suppose that could be a new step 9b that runs on all 4 machines.

Hopefully we can soon fully automate this.

----

Note I have a lot going on until June 1st so I might not be so active on this until then.

----

If you are interested, another big thing YS needs is package manager installs, like brew and apt and yum and on and on.

If you have any interest/time/expertise helping there would be awesome.

ejohb committed:
Add custom build_ext class for platform-specific wheel tagging

Introduce `LibYAMLScriptExtensionBuilder` to ensure setuptools correctly tags wheels with platform/architecture-specific metadata.

ejohb committed:
Revert test changes to Makefile and refine

ejohb committed:
Better handle shared library and package data in setup.py

ejohb committed:
Avoid building platform specific package if not required

ejohb committed:
Add show_info console script

ejohb committed:
Add test Dockerfile

ejohb commented:
OK, I'm now at the first stage I mentioned:

- Binary wheels are built for whichever of the supported platforms `make dist` is run under.
- I've tested this under `linux/amd64` and `linux/arm64`. See the release assets below.
- It should also work under macOS, but I don't have easy access to test.
- I've added the Dockerfile I used to test to the repo.
- I added a `show_info` console script for my own convenience when testing.

It would be cool if I could get the following feedback:

- Generally do a review let me know what you think.
- When it comes to modifying your flow so wheels are released for all platforms, let me know if you want to do that in a separate PR, or in this one.


[yamlscript-0.1.96-cp313-cp313-linux_x86_64.whl.gz](https://github.com/user-attachments/files/20493728/yamlscript-0.1.96-cp313-cp313-linux_x86_64.whl.gz)
[yamlscript-0.1.96-cp313-cp313-linux_aarch64.whl.gz](https://github.com/user-attachments/files/20493732/yamlscript-0.1.96-cp313-cp313-linux_aarch64.whl.gz)
[bdist_wheel.info.txt](https://github.com/user-attachments/files/20493736/bdist_wheel.info.txt)
[sdist.info.txt](https://github.com/user-attachments/files/20493737/sdist.info.txt)
[yamlscript-0.1.96.tar.gz](https://github.com/user-attachments/files/20493738/yamlscript-0.1.96.tar.gz)

ejohb commented:
Yep no worries. Just to clarify that the current code should work with the full matrix now, albeit if you run `make dist` manually. It would be good if you could do so actually (whether during a dry-run or not) just so we can have some confirmation the wheel built correctly under macOS.

ingydotnet commented:
I reworked the entire YS release system and added Python wheel support.
